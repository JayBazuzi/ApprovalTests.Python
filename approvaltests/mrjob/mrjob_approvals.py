from io import BytesIO
from itertools import product

from mrjob.job import MRJob
from approvaltests import verify, verify_all_combinations


def verify_map_reduce(mr_job_under_test: MRJob, test_data: str) -> None:
    storyboard = print_map_reduce_job(mr_job_under_test, test_data)
    verify(storyboard)


def print_map_reduce_job(mr_job_under_test: MRJob, test_data: str) -> str:
    storyboard = f"{test_data}\n\nMap reduces to:\n\n"
    mr_job_under_test.sandbox(stdin=BytesIO(test_data.encode("utf-8")))
    with mr_job_under_test.make_runner() as runner:
        runner.run()
        for key, value in mr_job_under_test.parse_output(runner.cat_output()):
            storyboard += f"{key}:{value}\n"
    return storyboard


def verify_templated_map_reduction(map_reduction, input_creator, params):
    inputs = product(*params)
    storyboard = ""
    for input in inputs:
        storyboard += f"===================\n\n{input} =>\n"
        data = input_creator(*input)
        storyboard += f"{print_map_reduce_job(map_reduction, data)}\n"
    verify(storyboard)
