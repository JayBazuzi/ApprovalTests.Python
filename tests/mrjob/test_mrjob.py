from mrjob.job import MRJob

from approvaltests.mrjob.mrjob_approvals import (
    verify_map_reduce,
    verify_templated_map_reduction,
)


class MRWordFrequencyCount(MRJob):
    def __init__(self, args=["--no-conf"]):
        super().__init__(args)

    def mapper(self, _, line):
        yield "chars", len(line)
        yield "words", len(line.split())
        yield "lines", 1

    def reducer(self, key, values):
        yield key, sum(values)


def test_word_count():
    test_data = "one fish two fish red fish blue fish"
    map_reduction = MRWordFrequencyCount()
    verify_map_reduce(map_reduction, test_data)


def test_word_count():
    test_data = "one fish two fish red fish blue fish"
    map_reduction = MRWordFrequencyCount()
    verify_map_reduce(map_reduction, test_data)


def test_word_count_combinations():
    animals = ["fish", "dog", "cat"]
    colors = ["aqua", "red", "blue"]
    map_reduction = MRWordFrequencyCount()

    def input_creator(animal, color1, color2):
        return f"one {animal} two {animal} {color1} {animal} {color2} {animal}"

    verify_templated_map_reduction(
        map_reduction, input_creator, [animals, colors, colors]
    )
