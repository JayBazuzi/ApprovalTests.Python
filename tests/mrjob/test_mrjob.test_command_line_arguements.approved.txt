===================

('aqua', 'cat') =>
one cat two cat red cat blue cat

Map reduces to:

aqua:1
chars:32
lines:1
words:8

===================

('aqua', 'dog') =>
one dog two dog red dog blue dog

Map reduces to:

aqua:1
chars:32
lines:1
words:8

===================

('blue', 'cat') =>
one cat two cat red cat blue cat

Map reduces to:

blue:1
chars:32
lines:1
words:8

===================

('blue', 'dog') =>
one dog two dog red dog blue dog

Map reduces to:

blue:1
chars:32
lines:1
words:8

